dao-modules/src/governance_voting.rs

```diff
@@ -0,0 +1,438 @@
+    consts, Address, Error, Mapping, Variable,
+};
+use casper_types::{runtime_args, RuntimeArgs, U256};
+
+use self::{
+    events::{
+        FormalVotingEnded, InformalVotingEnded, VoteCast, VotingContractCreated, VotingCreated,
+    },
+    vote::Vote,
+    voting::{Voting, VotingId},
+};
+
+/// The Governance Voting module.
+
+pub struct GovernanceVoting {
+    pub variable_repo: Variable<Option<Address>>,
+    pub reputation_token: Variable<Option<Address>>,
+    pub votings: Mapping<U256, Voting>,
+    pub votes: Mapping<(U256, Address), Vote>,
+    pub voters: Mapping<U256, Vec<Option<Address>>>,
+    pub votings_count: Variable<U256>,
+    pub dust_amount: Variable<U256>,
+}
+
+impl Default for GovernanceVoting {
+    fn default() -> Self {
+        Self {
+            variable_repo: Variable::from(consts::NAME_VARIABLE_REPO),
+            reputation_token: Variable::from(consts::NAME_REPUTATION_TOKEN),
+            votings: Mapping::from(consts::NAME_VOTINGS),
+            votes: Mapping::from(consts::NAME_VOTES),
+            voters: Mapping::from(consts::NAME_VOTERS),
+            votings_count: Variable::from(consts::NAME_VOTINGS_COUNT),
+            dust_amount: Variable::from(consts::NAME_DUST_AMOUNT),
+        }
+    }
+}
+
+impl GovernanceVoting {
+    /// Initialize the module.
+    pub fn init(&mut self, variable_repo: Address, reputation_token: Address) {
+        self.variable_repo.set(Some(variable_repo));
+        self.reputation_token.set(Some(reputation_token));
+        // Emit the event
+        let event = VotingContractCreated {
+            variable_repo,
+            reputation_token,
+            repo_voter: self_address(),
+        };
+
+        emit(event);
+    }
+
+    pub fn get_dust_amount(&self) -> U256 {
+        self.dust_amount.get()
+    }
+
+    pub fn get_variable_repo_address(&self) -> Address {
+        self.variable_repo.get().unwrap_or_revert()
+    }
+
+    pub fn create_voting(&mut self, voting: &Voting, stake: U256) {
+        // Add Voting
+        self.votings.set(&voting.voting_id, voting.clone());
+        self.votings_count.set(voting.voting_id + 1);
+
+        // Emit the event
+        let event = VotingCreated {
+            creator: caller(),
+            voting_id: voting.voting_id,
+            stake,
+        };
+
+        emit(event);
+
+        // Cast first vote in favor
+        self.vote(voting.voting_id, true, stake);
+    }
+
+    pub fn finish_voting(&mut self, voting_id: VotingId) {
+        let voting = self.votings.get(&voting_id);
+
+        // we cannot finish already completed voting
+        if voting.completed {
+            revert(Error::FinishingCompletedVotingNotAllowed)
+        }
+
+        // is it formal or informal?
+        if voting.voting_id == voting.informal_voting_id {
+            self.finish_informal_voting(voting);
+        } else if voting.voting_id == voting.formal_voting_id.unwrap() {
```

Comment:
Wouldn't it fail if `formal_voting_id` is  None? This `unwrap()` looking unsafe.
```suggestion
        } else if Some(voting.voting_id) == voting.formal_voting_id {
```

Reply:
Moved code to Voting and set to your proposal.